packages/cozy-client/src/helpers.js

```diff
@@ -75,3 +75,108 @@ export const generateWebLink = ({
 
   return url.toString()
 }
+
+const uri = (
+  {
+    protocol,
+    hostname,
+    port
+  } /*: { protocol: string, hostname: string, port: string } */
+) /*: string */ => {
+  return (
+    (protocol !== '' ? `${protocol}//` : 'https://') +
+    hostname +
+    (port !== '' ? `:${port}` : '')
+  )
+}
+
+export class InvalidProtocolError extends Error {
+  constructor(url) {
+    super(`Invalid URL protocol ${url.protocol}`)
+
+    this.url = url
+  }
+}
+
+export class InvalidCozyUrlError extends Error {
+  constructor(url) {
+    super(`URL ${url.toString()} does not seem to be a valid Cozy URL`)
+
+    this.url = url
+  }
+}
+
+const wellKnownUrl = (
+  url /*: { protocol: string, hostname: string, port: string } */
+) => uri(url) + '/.well-known/change-password'
+
+const isValidOrigin = async (
+  url /*: { protocol: string, hostname: string, port: string } */
+) /*: Promise<boolean> */ => {
+  const { status } = await fetch(wellKnownUrl(url))
+
+  if (status === 200) {
+    return true
+  } else if (status === 404) {
+    throw new InvalidCozyUrlError(url)
+  } else {
+    return false
+  }
+}
+
+/**
+ * rootCozyUrl - Get the root URL of a Cozy from more precise ones
+ *
+ * The goal is to allow users to use any URL copied from their browser as their
+ * Cozy URL rather than trying to explain to them what we expect (e.g. when
+ * requesting the Cozy URL to connect an app).
+ * If we can't get the root URL either because there's no Cozy or the domain
+ * does not exist or anything else, we'll throw an InvalidCozyUrlError.
+ * Also, since we communicate only via HTTP or HTTPS, we'll throw an
+ * InvalidProtocolError if any other protocol is used.
+ *
+ * @param {URL} url The URL from which we'll try to get the root Cozy URL
+ *
+ * @returns {Promise<URL>} The root Cozy URL
+ */
+export const rootCozyUrl = async url => {
+  if (!['http:', 'https:'].includes(url.protocol)) {
+    throw new InvalidProtocolError(url)
+  }
+
+  // If the entered URL is good, use it
+  if (await isValidOrigin(url)) {
+    return url
+  }
+
+  // If the entered URL's lowest sub-domain contains a dash, remove it and
+  // what follows and try the new resulting url.
+  if (/^[^.]+-[^.]+\./.test(url.hostname)) {
+    // XXX: url.hostname always contains a `.` as we append `.mycozy.cloud`
```

Comment:
i see a polynomial regex error above this line. Why?

Reply:
I'm don't know what this issue is but I need to make a little tweak to the regexp. Maybe the error will go away with it.